pkg/credentialprovider/config.go

```diff
@@ -73,11 +73,15 @@ func GetPreferredDockercfgPath() string {
 	defer preferredPathLock.Unlock()
 	return preferredPath
 }
+func GetDockercfgPaths() []string {
+	return []string{GetPreferredDockercfgPath(), workingDirPath, homeDirPath, rootDirPath, homeJsonDirPath, rootJsonDirPath}
+}
 
-func ReadDockerConfigFile() (cfg DockerConfig, err error) {
-	// Try happy path first - latest config file
-	dockerConfigJsonLocations := []string{GetPreferredDockercfgPath(), workingDirPath, homeJsonDirPath, rootJsonDirPath}
-	for _, configPath := range dockerConfigJsonLocations {
+func ReadDockerConfigFile(paths []string) (cfg DockerConfig, err error) {
+	if len(paths) == 0 {
+		paths = GetDockercfgPaths()
```

Comment:
sorry, this isn't what I meant.

currently, in master, we iterate over two distinct sets of file paths, one for possible jsonconfig files, then one for old config files.

I don't think we should combine those into a single list... what if an old config file unmarshalled into a jsonconfig struct without erroring, but also dropped data in the process?

decomposing into functions like this would make them reusable externally without changing the current behavior at all:
```
// ReadDockerConfigFile delegates to ReadDockerConfigJSONFile() and ReadDockercfgFile() using the default search paths
ReadDockerConfigFile() (cfg DockerConfig, err error)

// ReadDockercfgFile attempts to read a legacy dockercfg file from the given paths.
// if searchPaths is empty, the default paths are used.
ReadDockercfgFile(searchPaths []string) (cfg DockerConfig, err error)
DefaultDockercfgPaths() []string

// DefaultDockerConfigJSONPaths attempts to read a docker config.json file from the given paths.
// if searchPaths is empty, the default paths are used.
DefaultDockerConfigJSONPaths() []string
ReadDockerConfigJSONFile(searchPaths []string) (cfg DockerConfig, err error)
```